common/canonical_serialization/README.md

```diff
@@ -0,0 +1,471 @@
+# Libra Canonical Serialization (LCS)
+
+## Overview
+
+This document defines Libra Canonical Serialization (LCS). LCS defines a deterministic means for translating a message or data structure into bytes irrespective of platform, architecture, or programming language.
+
+## Background
+
+In Libra, participants pass around messages or data structures that often times need to be signed by one party and verified by many. Serialization is the process of converting a message into a byte array. Most serialization approaches support loose standards such that two implementations can produce two different byte streams that would represent the identical message. As a result, messages that are passed between participants must either retain the original serialized bytes or risk losing the ability to verify the message. This creates a burden programmatically by requiring the code to maintain a copy of the bytes after deserializing the object often leads to confusion about the safety and correctness of this approach. While there exist a handful of existing deterministic serialization formats, e.g. XDR, there exists no mainstream, well maintained, and available deterministic serialization. To address this, we propose Libra Canonical Serialization that defines a deterministic means for translating a message into bytes.
+
+## Specification
+
+LCS supports the following primitive types:
+
+* Booleans
+* Unsigned 8-bit integers
+* Unsigned 16-bit integers
+* Unsigned 32-bit integers
+* Unsigned 64-bit integers
+* Length prefixed byte array
+* Fixed-size byte array
+* UTF-8 Encoded Strings
+* Structures
+
+## General structure
+
+The serialized format of a message does not self-specify its encoding, in other words LCS employs stream encoding enabling storage of arbitrary and variable length messages. LCS defines the structure of primitives like integers, booleans, and byte arrays. It uses a composition of these primitives to support advanced data structures such as enumerations (unions), maps, and objects. Because of this typeless but expressive nature, each message requires both a serializer and reciprocal deserializer.
+
+Unless specified, all numbers are stored in little endian format.
+
+### Booleans and Integers
+
+|Type	|Original data	|Hex representation	|Serialized format	|
```

Comment:
Indentation lines up poorly here.  Is there a way to get both the text and the markdown view to look nice?

Reply:
Ahh, have to be more thorough when copying from quip to text.